Fix backend not running in Docker

**Describe the bug**

```
bookapp_1  | Error: Unable to initialize main class com.karankumar.bookproject.BookProjectApplication
bookapp_1  | Caused by: java.lang.NoClassDefFoundError: org/modelmapper/ModelMapper
```

When I run `docker-compose up`, the database starts fine but the backend is not able to start because it cannot find model mapper.

**To Reproduce**
Steps to reproduce the behaviour:
1. Go to the backend/book-app directory
2. Run docker-compose up
3. See error

**Expected behaviour**

Backend container runs.

**Additional context**

- Read our contributing guidelines first!
- Discuss your implementation approach with us first
- Provide us with an estimate on how long you think it will take you to complete
- Branch from main and set main as the destination branch in your pull request

Add any other context about the problem here (e.g. stack trace, branch name, commit hash, etc.).

--- a/README.md
+++ b/README.md
@@ -47,11 +47,13 @@ Next, start Docker engine (Linux) or Docker desktop (macOS or Windows). If you'r
 
 You may want to also want to run our [Books API](https://github.com/Project-Books/books-api) to avoid seeing an error on the search page.
 
-## Running the database
+## Running the backend and database
 
-1. Navigate to the [backend/](https://github.com/Project-Books/book-project/tree/main/backend) directory in a terminal
-1. Start the PostgreSQL database using `docker-compose up -d db`
-   - May need to add sudo to this command
+1. Build the Docker images in the `backend/` directory using `docker-compose build`
+1. Start the backend and database containers using `docker-compose up` 
+1. When finished, run `docker-compose down` to stop and remove the containers
+
+If you are contributing to the backend, please ensure you run the unit tests manually (we supply the `-DskipTests` flag with Docker by default for convenience).
 
 ### Access database (optional)
 
@@ -66,14 +68,6 @@ For example, in DataGrip or IntelliJ Ultimate:
 
 ![image](https://user-images.githubusercontent.com/11173328/153755219-051627c5-f052-4db9-a223-091acb4b2e76.png)
 
-## Running the backend
-
-**Note:** The backend depends on the database, so please ensure you have started our MySQL database with the instructions in the [section above](https://github.com/Project-Books/book-project#running-the-database).
-
-1. Build the Docker images in the `backend/` directory using `docker-compose build`
-1. Start the containers using `docker-compose up` 
-
-If you are contributing to the backend, please ensure you run the unit tests manually (we supply the `-DskipTests` flag with Docker by default for convenience).
 
 ## Log in with our test user
 

--- a/README.md
+++ b/README.md
@@ -73,6 +73,8 @@ For example, in DataGrip or IntelliJ Ultimate:
 1. Build the Docker images in the `backend/` directory using `docker-compose build`
 1. Start the containers using `docker-compose up` 
 
+If you are contributing to the backend, please ensure you run the unit tests manually (we supply the `-DskipTests` flag with Docker by default for convenience).
+
 ## Log in with our test user
 
 When running the frontend and backend, or only the backend, you can use the following test user:

--- a/README.md
+++ b/README.md
@@ -66,7 +66,7 @@ If you find this runs slowly (takes up too much of your computer's resources), y
 Using your favourite SQL client, use the following settings:
 - Host: `localhost`
 - Port: `5433`
-- User: `dbus
+- User: `dbuser`
 - Password: `dbpassword`
 - Database name: `book_project_db`
 

--- a/README.md
+++ b/README.md
@@ -53,7 +53,12 @@ You may want to also want to run our [Books API](https://github.com/Project-Book
 1. Start the backend and database containers using `docker-compose up` 
 1. When finished, run `docker-compose down` to stop and remove the containers
 
-If you are contributing to the backend, please ensure you run the unit tests manually (we supply the `-DskipTests` flag with Docker by default for convenience).
+If you find this runs slowly (takes up too much of your computer's resources), you can try:
+- Start the PostgreSQL database inside a container (`docker-compose up db`) but run the backend as a Maven app in your IDE
+   - note: if asking us for help, please ensure you use Docker
+- Using something more lightweight than an IDE like IntelliJ, such as VS Code
+
+> If you are contributing to the backend, please ensure you run the unit tests manually (we supply the `-DskipTests` flag with Docker by default for convenience).
 
 ### Access database (optional)
 
@@ -75,7 +80,7 @@ When running the frontend and backend, or only the backend, you can use the foll
 - Email address: `user@user.user`
 - Password: `password`
 
-If you're running the backend, you will need a JWT token for subsequent requests after logging in or creating an account; see our [connecting to the backend](https://github.com/Project-Books/book-project/wiki/Connecting-to-the-backend-via-Postman) wiki page.
+Note: If you're running the backend, you will need a JWT token for subsequent requests after logging in or creating an account; see our [connecting to the backend](https://github.com/Project-Books/book-project/wiki/Connecting-to-the-backend-via-Postman) wiki page.
 
 # Contributing
 

--- a/README.md
+++ b/README.md
@@ -29,7 +29,6 @@ Book tracker web app made with Spring Boot and React (Typescript).
 # Getting started locally
 
 Prerequisites:
-- JDK 11
 - Docker
   - Windows or macOS: install [Docker Desktop](https://www.docker.com/products/docker-desktop)
   - Linux: install [Docker Engine](https://docs.docker.com/engine/) and [Docker Compose](https://docs.docker.com/compose/)
@@ -71,10 +70,8 @@ For example, in DataGrip or IntelliJ Ultimate:
 
 **Note:** The backend depends on the database, so please ensure you have started our MySQL database with the instructions in the [section above](https://github.com/Project-Books/book-project#running-the-database).
 
-You can either run [BookProjectApplication.java](https://github.com/Project-Books/book-project/blob/main/backend/book-app/src/main/java/com/karankumar/bookproject/BookProjectApplication.java#L39) from your IDE (import as a Maven project) or run the steps below in a terminal:
-
-1. Build the project in the `backend/` directory using `./mvnw clean install` (Unix) or `mvnw.cmd clean install` (Windows)
-1. Start the application using `java -jar book-app/target/book-app-0.2.0.jar` 
+1. Build the Docker images in the `backend/` directory using `docker-compose build`
+1. Start the containers using `docker-compose up` 
 
 ## Log in with our test user
 

--- a/README.md
+++ b/README.md
@@ -29,7 +29,7 @@ Book tracker web app made with Spring Boot and React (Typescript).
 # Getting started locally
 
 Prerequisites:
-- Docker
+- Docker with [Buildkit enabled](https://docs.docker.com/develop/develop-images/build_enhancements/#to-enable-buildkit-builds)
   - Windows or macOS: install [Docker Desktop](https://www.docker.com/products/docker-desktop)
   - Linux: install [Docker Engine](https://docs.docker.com/engine/) and [Docker Compose](https://docs.docker.com/compose/)
 

--- a/README.md
+++ b/README.md
@@ -42,6 +42,7 @@ Next, start Docker engine (Linux) or Docker desktop (macOS or Windows). If you'r
 1. Navigate to the [frontend/](https://github.com/Project-Books/book-project/tree/main/frontend) directory in a terminal
 1. Run `docker-compose up --build` to build and start the frontend container
 1. Open `localhost:3000` once the development server has started
+1. When finished, run `docker-compose down` to stop and remove the frontend container
 
 ### Run Books API (optional)
 
@@ -54,18 +55,18 @@ You may want to also want to run our [Books API](https://github.com/Project-Book
 1. When finished, run `docker-compose down` to stop and remove the containers
 
 If you find this runs slowly (takes up too much of your computer's resources), you can try:
-- Start the PostgreSQL database inside a container (`docker-compose up db`) but run the backend as a Maven app in your IDE
+- Start the PostgreSQL database inside a container (`docker-compose up db`) but run the backend as a Maven app in your IDE. You will need to uncomment and comment the relevant lines in application-dev.properties [INSERT LINK]
    - note: if asking us for help, please ensure you use Docker
 - Using something more lightweight than an IDE like IntelliJ, such as VS Code
 
-> If you are contributing to the backend, please ensure you run the unit tests manually (we supply the `-DskipTests` flag with Docker by default for convenience).
+> Note for backend contributors: Please ensure you run the unit tests manually (we supply the `-DskipTests` flag with Docker by default for convenience).
 
 ### Access database (optional)
 
 Using your favourite SQL client, use the following settings:
 - Host: `localhost`
 - Port: `5433`
-- User: `dbuser`
+- User: `dbus
 - Password: `dbpassword`
 - Database name: `book_project_db`
 

--- a/backend/.dockerignore
+++ b/backend/.dockerignore
@@ -1,3 +1,7 @@
-#*
-#!target/book-project*.jar
+# IntelliJ
+.idea
+ide
 
+**/target/
+
+**/.md

--- a/backend/.dockerignore
+++ b/backend/.dockerignore
@@ -1,3 +1,3 @@
-*
-!target/book-project*.jar
+#*
+#!target/book-project*.jar
 

--- a/backend/Dockerfile
+++ b/backend/Dockerfile
@@ -1,0 +1,22 @@
+FROM maven:3.8.3-openjdk-11-slim AS builder
+WORKDIR /app
+COPY . /app
+ARG mvn_arg="clean package -DskipTests"
+
+RUN --mount=type=cache,target=/root/.m2 mvn -f /app/pom.xml $mvn_arg
+
+#
+
+FROM openjdk:11-jre-slim
+
+ARG JAR_FILE=/app/book-app/target/book-app-*.jar
+
+
+COPY --from=builder $JAR_FILE /opt/bookapp/app.jar
+
+WORKDIR /opt/bookapp/
+
+RUN useradd -m myuser
+USER myuser
+EXPOSE 8080
+CMD ["java", "-jar", "-Dspring.profiles.active=dev", "app.jar"]

--- a/backend/Dockerfile
+++ b/backend/Dockerfile
@@ -1,3 +1,4 @@
+# Create jar
 FROM maven:3.8.3-openjdk-11-slim AS builder
 WORKDIR /app
 COPY . /app
@@ -5,15 +6,10 @@ ARG mvn_arg="clean package -DskipTests"
 
 RUN --mount=type=cache,target=/root/.m2 mvn -f /app/pom.xml $mvn_arg
 
-#
-
+# Run jar
 FROM openjdk:11-jre-slim
-
 ARG JAR_FILE=/app/book-app/target/book-app-*.jar
-
-
 COPY --from=builder $JAR_FILE /opt/bookapp/app.jar
-
 WORKDIR /opt/bookapp/
 
 RUN useradd -m myuser

--- a/backend/book-app/Dockerfile
+++ b/backend/book-app/Dockerfile
@@ -1,9 +1,0 @@
-FROM openjdk:11-jre-slim
-ARG DEFAULT_PATH=target/book-app-*.jar
-WORKDIR /
-ENV JAR_PATH=$DEFAULT_PATH
-COPY ${JAR_PATH} app.jar
-RUN useradd -m myuser
-USER myuser
-EXPOSE 8080
-CMD ["java", "-jar", "-Dspring.profiles.active=dev", "app.jar"]

--- a/backend/book-app/pom.xml
+++ b/backend/book-app/pom.xml
@@ -6,6 +6,10 @@
   <build>
     <plugins>
       <plugin>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-maven-plugin</artifactId>
+      </plugin>
+      <plugin>
         <groupId>org.apache.maven.plugins</groupId>
         <artifactId>maven-compiler-plugin</artifactId>
         <configuration>

--- a/backend/book-app/src/main/resources/application-dev.properties
+++ b/backend/book-app/src/main/resources/application-dev.properties
@@ -3,7 +3,9 @@ spring.banner.image.location=banner.txt
 spring.jpa.hibernate.ddl-auto=validate
 spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.PostgreSQLDialect
 
+# Uncomment the line below if not running the backend in Docker:
 #spring.datasource.url = jdbc:postgresql://localhost:5433/book_project_db
+# comment out the line below if not running the backend in Docker:
 spring.datasource.url = jdbc:postgresql://book_project_db:5432/book_project_db
 spring.datasource.username = dbuser
 spring.datasource.password = dbpassword

--- a/backend/book-app/src/main/resources/application-dev.properties
+++ b/backend/book-app/src/main/resources/application-dev.properties
@@ -3,7 +3,8 @@ spring.banner.image.location=banner.txt
 spring.jpa.hibernate.ddl-auto=validate
 spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.PostgreSQLDialect
 
-spring.datasource.url = jdbc:postgresql://localhost:5433/book_project_db
+#spring.datasource.url = jdbc:postgresql://localhost:5433/book_project_db
+spring.datasource.url = jdbc:postgresql://book_project_db:5432/book_project_db
 spring.datasource.username = dbuser
 spring.datasource.password = dbpassword
 spring.datasource.driver-class-name=org.postgresql.Driver

--- a/backend/docker-compose.yml
+++ b/backend/docker-compose.yml
@@ -12,7 +12,6 @@ services:
     ports:
       - "5433:5432"
   bookapp:
-#    build: book-app
     container_name: book_project_backend
     build:
       context: .

--- a/backend/docker-compose.yml
+++ b/backend/docker-compose.yml
@@ -12,8 +12,14 @@ services:
     ports:
       - "5433:5432"
   bookapp:
-    build: ./
-    restart: on-failure      
+#    build: book-app
+    container_name: book_project_backend
+    build:
+      context: .
+      dockerfile: ./Dockerfile
+      args:
+        mvn_arg: clean package -DskipTests
+    restart: "no"
     ports: 
       - "8080:8080"
     environment:

--- a/backend/docker-compose.yml
+++ b/backend/docker-compose.yml
@@ -27,4 +27,6 @@ services:
       - POSTGRES_PASSWORD=dbpassword
       - POSTGRES_HOSTNAME=db
       - POSTGRES_PORT=5433
+    depends_on:
+      - db
 

--- a/backend/pom.xml
+++ b/backend/pom.xml
@@ -183,7 +183,6 @@
     </dependencies>
 
     <build>
-<!--        <defaultGoal>spring-boot:run</defaultGoal>-->
         <defaultGoal>clean package</defaultGoal>
         <pluginManagement>
             <plugins>
@@ -274,22 +273,6 @@
                     </execution>
                 </executions>
             </plugin>
-<!--            <plugin>-->
-<!--                <groupId>org.springframework.boot</groupId>-->
-<!--                <artifactId>spring-boot-maven-plugin</artifactId>-->
-<!--                <configuration>-->
-<!--                    <wait>500</wait>-->
-<!--                    <maxAttempts>240</maxAttempts>-->
-<!--                </configuration>-->
-<!--                <executions>-->
-<!--                    <execution>-->
-<!--                        <id>repackage</id>-->
-<!--                        <configuration>-->
-<!--                            <classifier>exec</classifier>-->
-<!--                        </configuration>-->
-<!--                    </execution>-->
-<!--                </executions>-->
-<!--            </plugin>-->
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-dependency-plugin</artifactId>

--- a/backend/pom.xml
+++ b/backend/pom.xml
@@ -183,7 +183,16 @@
     </dependencies>
 
     <build>
-        <defaultGoal>spring-boot:run</defaultGoal>
+<!--        <defaultGoal>spring-boot:run</defaultGoal>-->
+        <defaultGoal>clean package</defaultGoal>
+        <pluginManagement>
+            <plugins>
+                <plugin>
+                    <groupId>org.springframework.boot</groupId>
+                    <artifactId>spring-boot-maven-plugin</artifactId>
+                </plugin>
+            </plugins>
+        </pluginManagement>
         <plugins>
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
@@ -265,22 +274,22 @@
                     </execution>
                 </executions>
             </plugin>
-            <plugin>
-                <groupId>org.springframework.boot</groupId>
-                <artifactId>spring-boot-maven-plugin</artifactId>
-                <configuration>
-                    <wait>500</wait>
-                    <maxAttempts>240</maxAttempts>
-                </configuration>
-                <executions>
-                    <execution>
-                        <id>repackage</id>
-                        <configuration>
-                            <classifier>exec</classifier>
-                        </configuration>
-                    </execution>
-                </executions>
-            </plugin>
+<!--            <plugin>-->
+<!--                <groupId>org.springframework.boot</groupId>-->
+<!--                <artifactId>spring-boot-maven-plugin</artifactId>-->
+<!--                <configuration>-->
+<!--                    <wait>500</wait>-->
+<!--                    <maxAttempts>240</maxAttempts>-->
+<!--                </configuration>-->
+<!--                <executions>-->
+<!--                    <execution>-->
+<!--                        <id>repackage</id>-->
+<!--                        <configuration>-->
+<!--                            <classifier>exec</classifier>-->
+<!--                        </configuration>-->
+<!--                    </execution>-->
+<!--                </executions>-->
+<!--            </plugin>-->
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-dependency-plugin</artifactId>

